src/provider/shplonk.rs

```diff
@@ -243,39 +239,25 @@ where
       return Err(NovaError::ProofVerifyError);
     }
 
-    // TODO:
-    // insert _P_of_x into every pi.evals_i[last]
-    // insert _C into pi.comms[0]
-    // compute commitment for eval and insert it into pi.comms[last]
+    // Check correlation between proof and commitment to original polynomial
+    let c_expected = pi.comms.first().unwrap();
+    if C.comm.to_affine() != *c_expected {
+      return Err(NovaError::ProofVerifyError);
+    }
 
     let q = HyperKZG::<E, NE>::get_batch_challenge(&pi.evals, transcript);
-    //let q_powers = HyperKZG::<E, NE>::batch_challenge_powers(q, pi.comms.len());
-
     let R_x = UniPoly::new(pi.R_x.clone());
 
-    let mut evals_at_r = vec![];
-    let mut evals_at_minus_r = vec![];
-    let mut evals_at_r_squared = vec![];
-    for (i, evals_i) in pi.evals.iter().enumerate() {
-      if i == 0 {
-        evals_at_r = evals_i.clone();
-      }
-      if i == 1 {
-        evals_at_minus_r = evals_i.clone();
-      }
-      if i == 2 {
-        evals_at_r_squared = evals_i.clone();
-      }
-
-      let batched_eval = UniPoly::ref_cast(evals_i).evaluate(&q);
-
+    let verification_failed = pi.evals.par_iter().enumerate().any(|(i, evals_i)| {
```

Comment:
Parallel iteration here may be slower due to the small sizes.

Reply:
Right. We know that size of iterators is only 3